### PR TITLE
[WIP] Migrate to Puppet 4 datatypes to get rid of validate_* functions

### PR DESCRIPTION
**Pull request by bastelfreak**
The goal of this is to get rid of all the validate_* calls because they are deprecated.

### CODE REVIEW

File: manifests/plugin/discovery.pp [hunk 1]
```diff
@@ -17,22 +17,18 @@
 #                    type:String
 #
 class foreman_proxy::plugin::discovery (
-  $install_images = $::foreman_proxy::plugin::discovery::params::install_images,
-  $tftp_root      = $::foreman_proxy::plugin::discovery::params::tftp_root,
-  $source_url     = $::foreman_proxy::plugin::discovery::params::source_url,
-  $image_name     = $::foreman_proxy::plugin::discovery::params::image_name,
+  Boolean $install_images         = $::foreman_proxy::plugin::discovery::params::install_images,
+  Stdlib::Absolutepath $tftp_root = $::foreman_proxy::plugin::discovery::params::tftp_root,
```
Comment (bastelfreak) [inline manifests/plugin/discovery.pp]: I had the same question in my mind. But the params.pp provides working default settings so in my opinion the datatype is okay here.

File: manifests/plugin/dynflow.pp [hunk 2]
```diff
@@ -25,18 +25,14 @@
...
-  $listen_on         = $::foreman_proxy::plugin::dynflow::params::listen_on,
-  $database_path     = $::foreman_proxy::plugin::dynflow::params::database_path,
-  $console_auth      = $::foreman_proxy::plugin::dynflow::params::console_auth,
-  $core_listen       = $::foreman_proxy::plugin::dynflow::params::core_listen,
-  $core_port         = $::foreman_proxy::plugin::dynflow::params::core_port,
+  Boolean $enabled                    = $::foreman_proxy::plugin::dynflow::params::enabled,
+  Foreman_proxy::ListenOn $listen_on  = $::foreman_proxy::plugin::dynflow::params::listen_on,
+  Stdlib::Absolutepath $database_path = $::foreman_proxy::plugin::dynflow::params::database_path,
+  Boolean $console_auth               = $::foreman_proxy::plugin::dynflow::params::console_auth,
+  $core_listen                        = $::foreman_proxy::plugin::dynflow::params::core_listen,
```
Comment (bastelfreak) [inline manifests/plugin/dynflow.pp]: fixed

File: manifests/plugin/monitoring.pp [hunk 3]
```diff
@@ -23,15 +23,12 @@
...
-  $enabled            = $::foreman_proxy::plugin::monitoring::params::enabled,
-  $group              = $::foreman_proxy::plugin::monitoring::params::group,
-  $listen_on          = $::foreman_proxy::plugin::monitoring::params::listen_on,
-  $provider           = $::foreman_proxy::plugin::monitoring::params::provider,
-  $version            = $::foreman_proxy::plugin::monitoring::params::version,
+  Boolean $enabled                    = $::foreman_proxy::plugin::monitoring::params::enabled,
+  Optional[String] $group             = $::foreman_proxy::plugin::monitoring::params::group,
+  Foreman_proxy::ListenOn $listen_on  = $::foreman_proxy::plugin::monitoring::params::listen_on,
+  String $provider                    = $::foreman_proxy::plugin::monitoring::params::provider,
+  $version                            = $::foreman_proxy::plugin::monitoring::params::version,
```
Comment (bastelfreak) [inline manifests/plugin/monitoring.pp]: fixed

File: manifests/plugin/monitoring/icinga2.pp [hunk 4]
```diff
@@ -27,22 +27,15 @@
 # $enabled::               Enable this plugin.
 #
 class foreman_proxy::plugin::monitoring::icinga2 (
```
Comment (bastelfreak) [inline manifests/plugin/monitoring/icinga2.pp]: fixed

File: manifests/plugin/openscap.pp [hunk 5]
```diff
@@ -39,24 +39,16 @@
...
-  $version                 = $::foreman_proxy::plugin::openscap::params::version,
-  $listen_on               = $::foreman_proxy::plugin::openscap::params::listen_on,
-  $openscap_send_log_file  = $::foreman_proxy::plugin::openscap::params::openscap_send_log_file,
-  $spooldir                = $::foreman_proxy::plugin::openscap::params::spooldir,
-  $contentdir              = $::foreman_proxy::plugin::openscap::params::contentdir,
-  $reportsdir              = $::foreman_proxy::plugin::openscap::params::reportsdir,
-  $failed_dir              = $::foreman_proxy::plugin::openscap::params::failed_dir,
+  Boolean $configure_openscap_repo              = $::foreman_proxy::plugin::openscap::params::configure_openscap_repo,
+  Boolean $enabled                              = $::foreman_proxy::plugin::openscap::params::enabled,
+  $version                                      = $::foreman_proxy::plugin::openscap::params::version,
```
Comment (bastelfreak) [inline manifests/plugin/openscap.pp]: fixed

File: manifests/plugin/remote_execution/ssh.pp [hunk 6]
```diff
@@ -25,22 +25,18 @@
...
-  $ssh_identity_file  = $::foreman_proxy::plugin::remote_execution::ssh::params::ssh_identity_file,
-  $ssh_keygen         = $::foreman_proxy::plugin::remote_execution::ssh::params::ssh_keygen,
-  $local_working_dir  = $::foreman_proxy::plugin::remote_execution::ssh::params::local_working_dir,
-  $remote_working_dir = $::foreman_proxy::plugin::remote_execution::ssh::params::remote_working_dir,
+  Boolean $enabled                          = $::foreman_proxy::plugin::remote_execution::ssh::params::enabled,
+  Foreman_proxy::ListenOn $listen_on        = $::foreman_proxy::plugin::remote_execution::ssh::params::listen_on,
+  Boolean $generate_keys                    = $::foreman_proxy::plugin::remote_execution::ssh::params::generate_keys,
+  Stdlib::Absolutepath $ssh_identity_dir    = $::foreman_proxy::plugin::remote_execution::ssh::params::ssh_identity_dir,
+  String $ssh_identity_file                 = $::foreman_proxy::plugin::remote_execution::ssh::params::ssh_identity_file,
+  $ssh_keygen                               = $::foreman_proxy::plugin::remote_execution::ssh::params::ssh_keygen,
```
Comment (bastelfreak) [inline manifests/plugin/remote_execution/ssh.pp]: fixed

File: spec/spec.opts [hunk 7]
```diff
@@ -1,6 +1,5 @@
 --format
 documentation
 --colour
---loadby
```
Comment (ekohl) [inline spec/spec.opts]: This is managed by modulesync so will be overwritten on the next run.

Comment (bastelfreak) [inline spec/spec.opts]: sorry, this was committed by accident, I will purge it again. I had to remove this line to use parallel_spec for local testing.

**Comment by ekohl** [comment]
Sort of what I was afraid of: there's already a conflict.

File: manifests/plugin/monitoring/icinga2.pp [hunk 8]
```diff
@@ -16,33 +16,26 @@
...
-  $api_user = $::foreman_proxy::plugin::monitoring::icinga2::params::api_user,
-  $api_usercert = $::foreman_proxy::plugin::monitoring::icinga2::params::api_usercert,
-  $api_userkey = $::foreman_proxy::plugin::monitoring::icinga2::params::api_userkey,
-  $api_password = $::foreman_proxy::plugin::monitoring::icinga2::params::api_password,
-  $verify_ssl = $::foreman_proxy::plugin::monitoring::icinga2::params::verify_ssl,
+  Boolean $enabled                  = $::foreman_proxy::plugin::monitoring::icinga2::params::enabled,
+  String $server                    = $::foreman_proxy::plugin::monitoring::icinga2::params::server,
+  Stdlib::Absolutepath $api_cacert  = $::foreman_proxy::plugin::monitoring::icinga2::params::api_cacert,
+  String $api_user                  = $::foreman_proxy::plugin::monitoring::icinga2::params::api_user,
+  String $api_usercert              = $::foreman_proxy::plugin::monitoring::icinga2::params::api_usercert,
```
Comment (domcleal) [inline manifests/plugin/monitoring/icinga2.pp]: absolute path

File: manifests/plugin/monitoring/icinga2.pp [hunk 9]
```diff
@@ -16,33 +16,26 @@
...
+  String $api_userkey               = $::foreman_proxy::plugin::monitoring::icinga2::params::api_userkey,
```
Comment (domcleal) [inline manifests/plugin/monitoring/icinga2.pp]: absolute path

**Comment by bastelfreak** [comment]
@domcleal can you please submit the PR? Then we don't need to do the work twice.

**Comment by domcleal** [comment]
Sure, I've opened #344. Thanks @bastelfreak!

**Comment by mmoll** [comment]
#344 got merged. Thanks anyway @bastelfreak!

**Pull request closed without merging**